src/RedisIPCMessenger.ts

```diff
@@ -0,0 +1,130 @@
+import Redis from 'ioredis';
+import IPCMessenger, {
+  Room,
+  Instance,
+  MessageCallback,
+  MessageTypes,
+  Message,
+} from './IPCMessenger';
+
+type ConstructorParams = {
+  instance: Instance;
+  redis?: Redis.RedisOptions,
+  expireTime?: number,
+  refreshTime?: number,
+}
+
+class RedisIPCMessenger implements IPCMessenger {
+  private readonly instance: Instance;
+  private readonly publisher: Redis.Redis;
+  private readonly subscriber: Redis.Redis;
+  private readonly expireTime: number;
+  private readonly refreshTime: number;
+  private subscriptions: Map<Room, MessageCallback>;
+  private hasSetupCallbacks: boolean;
+
+  constructor(params: ConstructorParams) {
+    this.validateRoomOrInstance(params.instance);
+
+    this.instance = params.instance;
+    this.subscriptions = new Map();
+    this.publisher = new Redis(params.redis);
+    this.subscriber = new Redis(params.redis);
+    this.expireTime = params.expireTime ?? 30000;
+    this.refreshTime = params.refreshTime ?? 10000;
+    this.hasSetupCallbacks = false;
+  }
+
+  async join(room: Room, callback: MessageCallback): Promise<void> {
+    this.validateRoomOrInstance(room);
+
+    this.subscriptions.set(room, callback);
+
+    const refreshKeyLoop = async () => {
+      await this.publisher.set(`${room}:${this.instance}`, '', 'EX', this.expireTime / 1000);
+      setTimeout(refreshKeyLoop, this.refreshTime);
+    };
+    await refreshKeyLoop();
+
+    await this.subscriber.config('SET', 'notify-keyspace-events', 'Kx');
+    await this.subscriber.psubscribe(`__keyspace@0__:${room}:*`);
+    await this.subscriber.subscribe(`${room}`);
+
+    this.setupCallbacksIfNecessary();
```

Comment:
Is this really necessary? Why don't we set them up in the constructor?

Reply:
The callbacks setup is now also configuring redis to emit events for expired keys which is an async operation. Otherwise, it could have.

Reply:
Well, it's kind of obfuscated this way - this really is setup that is necessary, but cannot be performed inside the constructor. This problem is usually solved by using a dedicated `init` method. But, I guess it's fine as it is as well.